### PR TITLE
[JENKINS-38711] Add UncaughtExceptionHandler to remoting related threads

### PR DESCRIPTION
**Pull request by ghost**
See [JENKINS-38711](https://issues.jenkins-ci.org/browse/JENKINS-38711).

### Proposed changelog entries

* Entry 1: Handle uncaught exceptions in Tcp Agent Listener threads to restart them when possible.

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@oleg-nenashev @reviewbybees

### CODE REVIEW

File: core/src/main/java/hudson/TcpSlaveAgentListener.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/TcpSlaveAgentListener.java]: So, it stops the Listener, and it never tries to recover without the Jenkins restart? Then 🐛 

I propose to do the following:
* After shutdown schedule a delayed async task (e.g. 5 seconds), which starts a new thread.
* Modify the SEVERE message and say that the thread will be rescheduled
* Print INFO message when the listener thread starts up

File: core/src/main/java/hudson/TcpSlaveAgentListener.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/TcpSlaveAgentListener.java]: Something is missing here, right?
Have you checked that AgentProtocol gets aborted correctly in this case?

**Review by oleg-nenashev — CHANGES_REQUESTED** [review]
TcpSlaveAgentListener's behavior does not get worse, but I feel it's really important to try to reschedule the thread

File: core/src/main/java/hudson/TcpSlaveAgentListener.java
Comment (rysteboe) [inline core/src/main/java/hudson/TcpSlaveAgentListener.java]: Bad copy paste on the error.

I don't see how to abort AgentProtocol?  The only cleanup that I see is the socket, and that cleanup is attempted.

File: core/src/main/java/hudson/TcpSlaveAgentListener.java
Comment (rysteboe) [inline core/src/main/java/hudson/TcpSlaveAgentListener.java]: Did part of this, but I don't know how to do the wait (other than a `sleep(5000)`) before restarting the thread.

**Review by ghost — COMMENTED** [review]
Couple of questions in my responses.

File: core/src/main/java/hudson/TcpSlaveAgentListener.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/TcpSlaveAgentListener.java]: There is `AperiodicWork` in Jenkins, which can be used to schedule the delayed tasks

**Comment by rysteboe** [comment]
Updated.

**Comment by rysteboe** [comment]
Updated, though not sure I used that API correctly.

File: core/src/main/java/hudson/TcpSlaveAgentListener.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/TcpSlaveAgentListener.java]: It should also abort and restart the TcpSlaveAgentListener thread, no? Or at least to make the listener to restart the handler via a callback

**Comment by rysteboe** [comment]
Addressed your concerns (I think), although a little bit jankily.

File: core/src/main/java/hudson/TcpSlaveAgentListener.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/TcpSlaveAgentListener.java]: It may cay a race condition, because the current socket may be still handling Socket when the parent thread restarts. It may happen if the Thread is delayed for 5 seconds, which is probably non-realistic.

Comment (oleg-nenashev) [reply core/src/main/java/hudson/TcpSlaveAgentListener.java]: Maybe there should be different order

File: core/src/main/java/hudson/TcpSlaveAgentListener.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/TcpSlaveAgentListener.java]: Mark as `@Restricted`? I am not sure if it should be a part of the public API

File: core/src/main/java/hudson/TcpSlaveAgentListener.java
Comment (oleg-nenashev) [inline core/src/main/java/hudson/TcpSlaveAgentListener.java]: Reschedule another attempt? If the Agent port is enabled, there is no valid use-case for the dead thread. It would also indicate that there is a pending problem if the thread is alive and rejects to stop

**Review by oleg-nenashev — COMMENTED** [review]
Mostly looks good. Some minor patches would be useful, maybe it would be also useful to ensure the logic really runs

File: core/src/main/java/hudson/TcpSlaveAgentListener.java
Comment (rysteboe) [reply core/src/main/java/hudson/TcpSlaveAgentListener.java]: Flipped the order.

File: core/src/main/java/hudson/TcpSlaveAgentListener.java
Comment (rysteboe) [reply core/src/main/java/hudson/TcpSlaveAgentListener.java]: I moved `isActive = false` to only happen if the thread is successfully restarted.  I think that means that it will automatically retry?

**Review by ghost — COMMENTED** [review]
Applied those changes.

**Comment by oleg-nenashev** [comment]
@rysteboe Maybe I misunderstand the code. `TcpSlaveAgentListenerRescheduler `is and **A**periodic work, so I do not see how it will be automatically rescheduled.

Anyway, the code is almost there. 👍

**Comment by rysteboe** [comment]
@oleg-nenashev the Javadoc for `AperodicWork` says that it it is still scheduled on an interval, the difference is that `AperodicWork` can have that interval changed.  

https://github.com/jenkinsci/jenkins/blob/08def67a18eee51de9f3f99bc2a792fee1c160e0/core/src/main/java/hudson/model/AperiodicWork.java
https://github.com/jenkinsci/jenkins/blob/08def67a18eee51de9f3f99bc2a792fee1c160e0/core/src/main/java/hudson/model/PeriodicWork.java

Am I misunderstanding?

**Review by oleg-nenashev — APPROVED** [review]
No, you are right. 🐝 
Let's see if we can deliver Remoting and core changes in the same release. Do you consider this change as LTS candidate?

**Comment by oleg-nenashev** [comment]
Remoting PR: https://github.com/jenkinsci/remoting/pull/187 . I am not sure if you're waiting for my reviews there @rysteboe

**Comment by rysteboe** [comment]
@oleg-nenashev yeah, there was a question in there that got buried I think.

I really don't like Github's code review interface :/

**Comment by rysteboe** [comment]
re: releases, do I need to wait to merge until remoting is ready then?

re: LTS, I would think so, but I don't know much about the Jenkins LTS process

**Comment by oleg-nenashev** [comment]
This PR can be merged independently, it will just require a more specific changelog. I assume @reviewbybees done anyway, let's give some time to others to review this change

**Comment by oleg-nenashev** [comment]
This PR is open for 24 days, nobody else seemed to be interested.
Merging as a Remoting maintainer

**Comment by oleg-nenashev** [comment]
Probably I have merged it too early. The case like #3012 applies here as well. If somebody disables the protocols while the restart task in the queue, the Agent Listener will start independently on the the new setting. 

* So the restart handler should also check with the current settings and probably just stop when the port is disabled.  
* Probably it should catch other differences just in case

@rysteboe @daniel-beck Sorry for the mess, I am going to revert it and reopen the PR

**Pull request merged**
